File: src/cs/every-base/every-base/every_base.cs
using System;
using System.ComponentModel.DataAnnotations;

namespace every_base
{
    public class Bases
    {
        static void Main(string[] args)
        {
            
        }

        static int[] array_rotation_left(int[] array, int n)
        {
            int len = array.Length;
            n = n % len;

            for (int i = 0; i < n; i++)
            {
                for (int k = 0; k < len-1; k++)
                    swap(ref array, k, k+1);
            }
            
            return array;
        }

        static int[] array_rotation_right(int[] array, int n)
        {
            int len = array.Length;
            n = n % len;

            for (int i = 0; i < n; i++)
            {
                for (int k = len-1; k > 0; k--)
                    swap(ref array, k, k-1);
            }
            
            return array;
        }
        
        static int my_atoi(string str)
        {
            if(str[0]=='-')
                return  - (int) my_atoui(str.Substring(1));
            return (int) my_atoui(str);
        }

        // We assume the string correct (only digits char)
        static uint my_atoui(string str)
        {
            uint res = 0;

            int len = str.Length;
            for (int i = 0; i < len; res=(uint) (str[i]-'0'+res*10), i++) ;

            return res;
        }

        static bool binsearch(int[] array, int x)
        {
            int left = 0;
            int right = array.Length;

            while (left < right)
            {
                int mid = (right + left) / 2;

                if (array[mid] == x)
                    return true;
                else if (array[mid] > x)
                    right = mid;
                else
                    left = mid + 1;
            }
            
            return false;
        }

        static void bubble_sort(ref int[] array)
        {
            int len = array.Length;

            bool changed = true;
            for (int i = 1; i < len-1 && changed; i++)
            {
                changed = false;
                
                for (int k = 0; k < len-i; k++)
                {
                    if (array[k] > array[k + 1])
                    {
                        swap(ref array, k, k + 1);
                        changed = true;
                    }
                }
            }
        }
        
        static int[] char_histogram(string str)
        {
            int[] res = new int[26];
            int len = str.Length;
            
            for (int i=0; i<len; i++)
            {
                if (str[i] >= 'a' && str[i] <= 'z')
                    res[str[i] - 'a'] += 1;
                else if (str[i] >= 'A' && str[i] <= 'Z')
                    res[str[i] - 'A'] += 1;
            }
            
            return res;
        }

        static int count_words(string str)
        {
            int res = 0;
            int len = str.Length;
            bool separated = true;

            for (int i = 0; i < len; i++)
            {
                if ((str[i] >= 'a' && str[i] <= 'z') 
                    || (str[i] >= 'A' && str[i] <= 'Z') 
                    || (str[i] >= '0' && str[i] <= '9'))
                {
                    if (separated)
                        res += 1;
                    separated = false;
                }
                else
                    separated = true;
            }

            return res;
        }

        static int digit_count(int n)
        {
            int res = 1;

            for (; n > 9; res++, n /= 10) ;
            
            return res;
        }

        static int divisor_sum(int n)
        {
            int sum = 1;

            int i;
            for (i = 2; i * i < n; i++)
            {
                if (n % i == 0)
                    sum += i + n / i;
            }

            if (i * i == n)
                sum += i;
            
            return sum;
        }

        static int facto(int n)
        {
            int res = 1;

            for (int i = 2; i < n; res *= i, i++) ;
            
            return res;
        }

        static int fibo(int n)
        {
            if (n <= 0)
                return 0;
            
            int f0 = 0;
            int f1 = 1;

            for (int i = 1; i < n; i++)
            {
                int tmp = f1;
                f1 += f0;
                f0 = tmp;
            }

            return f1;
        }

        static void fizzbuzz(int n)
        {
            for (int i = 0; i <= n; i++)
            {
                bool div3;

                if ((div3 = (i%3==0)))
                    Console.Write("Fizz");
                if (i%5==0)
                    Console.Write("Buzz");
                else if (!div3)
                    Console.Write(i);
                    
                Console.WriteLine();
            }
        }

        static bool is_prime(int n)
        {
            if (n % 2 == 0 && n != 2)
                return false;
            
            for (int i = 3; i * i <= n; i += 2)
            {
                if (n % i == 0)
                    return false;
            }
            
            return true;
        }

        static int isqrt(int n)
        {
            int root = n;

            for (; root * root > n; root = ((root+ n)/root)/2 ) ;
            
            return root; 
        }

        static bool is_sorted(int[] array)
        {
            int len = array.Length;

            int i;
            for (i = 1; i < len && array[i] >= array[i - 1]; i++) ;
            
            return i>=len;
        }

        static int[] merge_sorted(int[] a1, int[] a2)
        {
            int len1 = a1.Length;
            int len2 = a2.Length;
            int[] res = new int[len1+len2];

            int i, j, k;
            i = j = k = 0;

            while (i < len1 || j < len2)
            {
                if (i < len1)
                {
                    if ((j < len2 && a1[i] <= a2[j]) || j >= len2)
                    {
                        res[k] = a1[i];
                        i++;
                        k++;
                    }
                }
                if (j < len2)
                {
                    if ((i < len1 && a2[j] <= a1[i]) || i >= len1)
                    {
                        res[k] = a2[j];
                        j++;
                        k++;
                    }
                }
            }

            return res;
        }

        static bool perfect_number(int n)
        {
            return divisor_sum(n) == n;
        }

        static int power(int x, int n)
        {
            if (x==1 || n==0)
                return 1;
            
            int res = 1;

            while( n > 0 )
            {
                if (n % 2 == 0)
                {
                    x *= x;
                    n /= 2;
                }
                else
                {
                    res *= x;
                    n--;
                }
            }

            return res;
        }

        static int power_of_two(int n)
        {
            return 1 << n;
        }

        static void quick_sort(ref int[] arr, int lo = 0, int hi = -2)
        {
            if (hi == -2)
                hi = arr.Length - 1;
            if (lo < hi)
            {
                int p = __partition(ref arr, hi, lo);
                
                quick_sort(ref arr, lo, p-1);
                quick_sort(ref arr, p+1, hi);
            }
        }

        static int __partition(ref int[] array, int lo, int hi)
        {
            int pivot = array[hi];
            int i = lo - 1;

            for (int j = lo; j <= hi; j++)
            {
                if (array[j] <= pivot)
                {
                    i++;
                    swap(ref array, i, j);
                }
            }

            return i;
        }

        static bool[] sieve_of_eratosthenes(int n)
        {
            bool[] res = new bool[n-1];
            
            for (int i = 0; i < n; res[i] = true, i++) ;
            
            for (int i = 0; (i + 2) * (i + 2) < n; i++)
            {
                if (res[i])
                {
                    for (int j = (i+2)*(i+2); j < n; j+=(i+2))
                        res[j - 2] = false;
                }
            }
            
            return res;
        }

        static void swap(ref int[] array, int i, int j)
        {
            int tmp = array[i];
            array[i] = array[j];
            array[j] = tmp;
        }

        static string to_lower(string str)
        {
            int len = str.Length;
            char shift = (char) ('a' - 'A');
            string res = "";

            for (int i = 0; i < len; i++)
            {
                if (str[i] >= 'A' && str[i] <= 'Z')
                    res += shift+str[i];
                else
                    res += str[i];
            }
            
            return res;
        }
        
        static string to_upper(string str)
        {
            int len = str.Length;
            char shift = unchecked((char) ('A' - 'a'));
            string res = "";

            for (int i = 0; i < len; i++)
            {
                if (str[i] >= 'a' && str[i] <= 'z')
                    res += shift+str[i];
                else
                    res += str[i];
            }
            
            return res;
        }
    }
}

File: src/java/every_base.java
package java;


public class every_base {
    
    public static int[] array_rot_left(int[] array, int n)
    {
        n = n % array.length;

        for (int i = 0; i < n; i++) {
            int moved = array[0];
            for (int j = array.length; j > 0; j--) {
                int tmp = array[j];
                array[j] = moved;
                moved = tmp;
            }

            array[0] = moved;
        }

        return array;
    }
    
    public static int[] array_rot_right(int[] array, int n)
    {
        n = n % array.length;

        for (int i = 0; i < n; i++) {
            int moved = array[array.length-1];
            for (int j = 0; j < array.length-1; j++) {
                int tmp = array[j];
                array[j] = moved;
                moved = tmp;
            }

            array[array.length-1] = moved;
        }

        return array;
    }

    public static int atoi(String str)
    {
        char[] char_arr = str.toCharArray();

        if(char_arr[0]=='-')
            return -atoui(str);
        return atoui(str);
    }

    public static int atoui(String str)
    {
        int result = 0;
        char[] char_arr = str.toCharArray();

        for (int i = 0; i < char_arr.length && (char_arr[i]>='0' && char_arr[i]<='9'); i++) 
            result = result*10 + char_arr[i]-'0';
        

        return result;
    }

    public static int[] bubble_sort(int[] array)
    {
        int n = array.length;

        while (n>1)
        {
            int newn = 0;
            for(int i = 1; i<n; i++)
            {
                if(array[i-1]>array[i])
                {
                    int tmp = array[i-1];
                    array[i-1] = array[i];
                    array[i] = tmp;
                    newn = i;
                }
                
            }
            n = newn;
        }

        return array;
    }

    public static boolean binary_search(int[] array, int searched)
    {
        int l = 0;
        int r = array.length-1;

        while (l<=r) {
            int mid = (l+r)/2;
            if(array[mid]==searched)
                return true;
            else if(array[mid]>searched)
                r = mid-1;
            else
                l = mid+1;
        }

        return false;
    }

    public static int[] char_histogram(String str)
    {
        int[] result = new int[26];

        char[] char_arr = str.toCharArray();

        for (int i = 0; i < char_arr.length; i++) {
            if(char_arr[i]>='a' && char_arr[i]<='z')
                result[i-'a']+=1;
            else if(char_arr[i]>='A' && char_arr[i]<='Z')
                result[i-'A']+=1;
        }

        return result;
    }

    public static int count_words(String str)
    {
        int count = 0;
        boolean isSeparated = true;

        for (int i = 0; i < str.length(); i++) 
        {
            char curr_char = str.charAt(i);
            if((curr_char >= 'a' && curr_char <= 'z') || 
                (curr_char >= 'A' && curr_char <= 'Z') || 
                (curr_char >= '0' && curr_char <= '9'))
            {
                if(isSeparated)
                {
                    isSeparated=false;
                    count++;
                }
            }
            else
                isSeparated = true;
        }

        return count;
    }

    public static char digit_count(long n)
    {
        char length = 0;

        do {
            length += 1;
            n /= 10;
        } while (n!=0);

        return length;
    }

    public static long divisor_sum(long n)
    {
        long sum = 0;

        long i = 2;
        int step = 1;

        if(n%2==0)
        {
            i = 3;
            step = 2;
        }

        for(; i*i<n; i+=step)
        {
            if(n%i == 0)
                sum += i + n/i;
        }

        return sum;
    }

    public static long facto(long n)
    {
        long result = 1;

        for(long i = 2; i<=n; result*=i, i++);

        return result;
    }

    public static long fibo(long n)
    {
        long f1 = 1;
        long f0 = 0;

        for(long i = 1; i<n; i++)
        {
            long tmp = f0;
            f0 = f1;
            f1 += tmp;
        }

        return f1;
    }

    public static void fizzbuzz(long n)
    {
        for (long i = 1; i <= n; i++) {
            Boolean isDiv3;
            if(isDiv3=i%3==0)
                System.out.print("Fizz");
            if(i%5==0)
                System.out.print("Buzz");
            else if(!isDiv3)
                System.out.print(i);

            System.out.println();
        }
        
    }

    public static boolean is_prime(long n)
    {
        if((n%2==0 && n!=2)|| n==1)
            return false;

        for(long i = 3; i*i<=n; i+=2)
        {
            if(n%i==0)
                return false;
        }

        return true;
    }

    public static long isqrt(long n)
    {
        long root = n;

        while(root*root>n)
            root = ((root+n)/root)/2;

        return root;
    }

    public static boolean is_sorted(int[] array)
    {
        for (int i = 0; i < array.length-1; i++) {
            if(array[i]>array[i+1])
                return false;
        }

        return true;
    }

    public static int[] merge_sorted(int[] array1, int[] array2)
    {
        int[] result = new int[array1.length+array2.length];

        int i, j, k;
        i = j = k = 0;

        while(i<array1.length || j<array2.length)
        {
            if(i<array1.length)
            {
                if(j>=array2.length || array1[i]<array2[j])
                {
                    result[k] = array1[i];
                    k++;
                    i++;
                }
            }
            if(j<array2.length)
            {
                if(i>=array1.length || array1[i]>array2[j])
                {
                    result[k] = array2[j];
                    k++;
                    j++;
                }
            }
        }

        return result;
    }

    public static boolean perfect_number(long n)
    {
        return divisor_sum(n)==n;
    }

    public static long power(long x, long n)
    {
        long result = 1;

        while(n>0)
        {
            if(n%2==1)
            {
                result *= x;
                n--;
            }
            else
            {
                n/=2;
                x*=x;
            }
        }

        return result;
    }

    public static long power_of_two(char n)
    {
        return 1<<n;
    }

    private static int[][] partition(int[] array, int low, int high)
    {
        int pivot = array[high];
        int i = low-1;

        for(int j = low; j<=high; j++)
        {
            if(array[j]<=pivot)
            {
                i++;
                array = swap(array, i, j);
            }
        }

        int[][] result = new int[2][];
        result[1] = new int[1];
        result[1][0] = i;
        result[0] = array;
        return result;
    }

    public static int[] quick_sort(int[] array, int low, int high)
    {
        if(low<high)
        {
            int[][] pivot = partition(array, low, high);

            pivot[0] = quick_sort(pivot[0], low, pivot[1][0]-1);
            pivot[0] = quick_sort(pivot[0], pivot[1][0]+1, high);

            array = pivot[0];
        }
        return array;
    }

    public static boolean[] sieve_of_eratosthenes(int n)
    {
        n+=1;
        boolean[] result = new boolean[n];
        for (int i = 0; i < result.length; i++) 
            result[i] = true;

        for (int i = 2; i*i < n; i++) {
            for(int k = i*i; k<n; k+=i)
                result[k] = false;
        }

        return result;
    }

    public static int[] swap(int[] array, int index1, int index2)
    {
        int tmp = array[index1];
        array[index1] = array[index2];
        array[index2] = tmp;

        return array;
    }

    
}
